components/experimental/src/dimension/provider/pattern_key.rs

```diff
@@ -0,0 +1,204 @@
+/// - `11`: Forbidden
+///
+/// The next 6 bits (b5 to b0) determine the value of the pattern key:
+/// - For `Binary`, the value is mapped directly to the pattern value.
+/// - For `Decimal`:
+///     - b5 is determining the sign of the value. if b5 is 0, the value is positive. if b5 is 1, the value is negative.
+///     - b4 to b0 are determining the magnitude of the value.
+/// - For `Power`:
+///     - b5 and b4 represent the power value, which can be `10` to represent `Two` and `11` to represent `Three`.
+///     - b3 to b0 represent the count value, which can be:
+///         - `0000`: Zero
+///         - `0001`: One
+///         - `0010`: Two
+///         - `0011`: Few
+///         - `0100`: Many
+///         - `0101`: Other
+///     - Note: In the `Power` case, b3 is always 0, and when b2 is 1, b1 must be 0.
+#[derive(Copy, Clone, PartialOrd, Ord, PartialEq, Eq, Debug)]
+pub struct PatternKeyULE(u8);
+
+unsafe impl ULE for PatternKeyULE {
+    fn validate_byte_slice(bytes: &[u8]) -> Result<(), zerovec::ZeroVecError> {
+        if bytes.len() != 1 {
+            return Err(ZeroVecError::length::<Self>(bytes.len()));
+        }
+
+        let byte = &bytes[0];
+
+        // Ensure the first two bits (b7 & b6) are not 11.
+        if (byte & 0b1100_0000) == 0b1100_0000 {
+            return Err(ZeroVecError::VarZeroVecFormatError);
+        }
+
+        // For the `Power` variant:
+        //      b5 & b4 must be 10 or 11. (this means that b5 must be 1)
+        //      b3 must be 0.
+        //      When b2 is 1, b1 must be 0.
+        if (byte & 0b1100_0000) == 0b1000_0000 {
+            // b5 must be 1
+            if (byte & 0b0010_0000) == 0 {
+                return Err(ZeroVecError::VarZeroVecFormatError);
+            }
+
+            // b3 must be 0
+            if (byte & 0b0000_1000) != 0 {
+                return Err(ZeroVecError::VarZeroVecFormatError);
+            }
+
+            // If b2 is 1, b1 must be 0
+            if (byte & 0b0000_0100) != 0 && (byte & 0b0000_0010) != 0 {
+                return Err(ZeroVecError::VarZeroVecFormatError);
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl AsULE for PatternKey {
+    type ULE = PatternKeyULE;
+
+    fn to_unaligned(self) -> Self::ULE {
+        let byte = match self {
+            PatternKey::Binary(value) => value,
+            PatternKey::Decimal(value) => {
+                // TODO: shall I check the limits of the values?
+                let sign = if value < 0 { 0b0010_0000 } else { 0 };
+                (0b01 << 6) + sign + (value as u8)
+            }
+            PatternKey::Power(power, count) => {
+                let power_bits = {
+                    match power {
+                        PowerValue::Two => 0b10 << 4,
+                        PowerValue::Three => 0b11 << 4,
+                    }
+                };
+                // Combine the bits to form the final byte
+                (0b10 << 6) + power_bits + count as u8
```

Comment:
Suggestion (optional):

```suggestion
                // Combine the bits to form the final byte
                (0b10 << 6) | power_bits | count as u8
```